brml/swap/src/lib.rs

```diff
@@ -1,1022 +1,1400 @@
+
+				let pool_token_out = Self::calculate_pool_in_given_single_out(out_token_weight, token_amount, 
+					out_token_balance_in_pool, pool_supply, swap_fee_rate)?;
+				let pool_token_out = u128::from_fixed(pool_token_out);
+
+				TryInto::<T::Balance>::try_into(pool_token_out).map_err(|_| Error::<T>::ConvertFailure)?
+			};
+
+			ensure!(UserPoolTokensInPool::<T>::get(&remover, pool_id) >= pool_token_out, Error::<T>::NotEnoughBalance);
+			ensure!(pool_token_out >= T::MinimumPassedInPoolTokenShares::get(), Error::<T>::LessThanMinimumPassedInPoolTokenShares);
+
+			// update user asset
+			T::AssetTrait::asset_issue(asset_id, &remover, token_amount);
+
+			// update TokenBalancesInPool map.
+			TokenBalancesInPool::<T>::mutate(pool_id, asset_id, |token_balance| {
+				*token_balance = token_balance.saturating_sub(token_amount);
+			});
+
+			Self::revise_storages_except_token_balances_when_removing_liquidity(pool_id, pool_token_out, &remover)?;
+			Self::deposit_event(RawEvent::RemoveSingleLiquiditySuccess);
+
+			Ok(())
+		}
+
+		// ****************************************************************************
+		/// User removes all the tokens in the pool in proportion of his pool token shares.
+		/// (remove liquidity)(many assets) given share in => amount out
+		#[weight = T::DbWeight::get().reads_writes(1, 1)]
+		fn remove_assets_liquidity_given_shares_in(
+			origin,
+			pool_id: T::PoolId,
+			#[compact] pool_amount_out: T::Balance
+		) {
+			let remover = ensure_signed(origin)?;
+
+			ensure!(Pools::<T>::contains_key(pool_id), Error::<T>::PoolNotExist);
+			ensure!(Pools::<T>::get(pool_id).active, Error::<T>::PoolNotActive);
+			ensure!(UserPoolTokensInPool::<T>::contains_key(&remover, pool_id), Error::<T>::UserNotInThePool);
+			ensure!(pool_amount_out >= T::MinimumPassedInPoolTokenShares::get(), Error::<T>::LessThanMinimumPassedInPoolTokenShares);
+
+			let user_pool_token_in_pool = UserPoolTokensInPool::<T>::get(remover.clone(), pool_id);
+			ensure!(pool_amount_out <= user_pool_token_in_pool, Error::<T>::NotEnoughBalance);
+			// get the iterator of the items(asset_id => balance) with the same first key(pool_id)
+			let token_balances_in_pool_iter = TokenBalancesInPool::<T>::iter_prefix(pool_id);
+			// calculate how many tokens for each of the assets that user can withdraw. 
+			// Meanwhile, issue money to user's account and deducted from the pool.
+			for tk in token_balances_in_pool_iter {  //0 position is asset_id, 1 position is balance
+				// get the total pool token shares for the specific pool
+				let all_pool_tokens = PoolTokensInPool::<T>::get(pool_id);
+				// Asset id
+				let token_id = tk.0;
+				// the balance of a specific token in a pool
+				let token_pool_balance = TokenBalancesInPool::<T>::get(pool_id, token_id);
+				// the amount of the token that the user should deposit
+				let can_withdraw_amount = token_pool_balance * pool_amount_out / all_pool_tokens;

```

Comment:
'*' shoule be replaced by saturating_mul

Reply:
fixed.